### PR TITLE
Aj issue2

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [x] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by adiengineer**
Closes #2

Finished up issue 2

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -4,19 +4,34 @@ class ImagesController < ApplicationController
...
     @image = Image.new
-
   end
 
   def create
-    @image = Image.create ( image_params )  #params[:image][:image_uri]  )
+    @image = Image.new(image_params)
+    if @image.valid?
+      # save image and redirect
+      @image.save
```
Comment (abineetds) [inline app/controllers/images_controller.rb]: It's usually better practice to use `save!` instead of `save`. The former throws an error if the save fails, the latter just returns a boolean value. The other thing you could do is rather than doing `if @image.valid?` you could do `if @image.save`. Though it might just be better to use `save!` here so that if, for instance the write to the database fails, it reports an error

File: test/models/image_test.rb [hunk 2]
```diff
@@ -4,4 +4,25 @@ class ImageTest < ActiveSupport::TestCase
   # test "the truth" do
```
Comment (abineetds) [inline test/models/image_test.rb]: you should remove auto-generated comments

File: test/models/image_test.rb [hunk 3]
```diff
@@ -7,22 +7,52 @@ class ImageTest < ActiveSupport::TestCase
   #
 
 
+  # the test function which is always called first when image tester is invoked
+  # rets: a sample test object
```
Comment (abineetds) [inline test/models/image_test.rb]: it's not really returning the sample test object. It's just setting the instance variable

File: test/models/image_test.rb [hunk 4]
```diff
@@ -7,22 +7,52 @@ class ImageTest < ActiveSupport::TestCase
...
   end
 
 
+  # take img obj to a wrong state and check if that error is caught
+  # this is prefered in appfolio
+
+  # testing style: cause that error and then assert if the expected error message is caught
   def test_is_url_filled
     @img.image_uri = ""
     refute_predicate(@img,  :valid?)
```
Comment (abineetds) [inline test/models/image_test.rb]: You don't need () here

File: test/models/image_test.rb [hunk 5]
```diff
@@ -7,22 +7,52 @@ class ImageTest < ActiveSupport::TestCase
...
-    assert_equal(["can't be blank"],@img.errors[:image_uri])
+
+    # every resour object is assoc with an error message hash- one for every attr?
+
+    # this checks if there was any err assoc with image_uri attr and compares it to the hard coded message we
+    # got from the sandbox
+    assert_equal(["can't be blank", "is invalid"],@img.errors[:image_uri])
```
Comment (abineetds) [inline test/models/image_test.rb]: You don't need () here either

File: .rubocop.yml [hunk 6]
```diff
@@ -6,6 +6,7 @@ AllCops:
     - 'vendor/**/*'
     - 'node_modules/**/*'
     - 'bin/*'
+    - '/Users/aditya.joglekar/RubymineProjects/ropes_dir/image-sharer-adiengineer/config/initializers/simple_form.rb'
```
Comment (abineetds) [inline .rubocop.yml]: You don't need to put the full path here. you can just put /config/....

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -18,26 +15,24 @@ def create
     else
       # re-render the form
       # the error message if any is nicely displayed on re-rendering, simple form takes care of that
-      render( :new )
+      render(:new)
```
Comment (abineetds) [inline app/controllers/images_controller.rb]: Dont need () here

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -18,26 +15,24 @@ def create
...
     end
 
-    #@image = Image.create ( image_params )  #params[:image][:image_uri]  )
-    #redirect_to action: "show"
+    # @image = Image.create ( image_params )  #params[:image][:image_uri]  )
```
Comment (abineetds) [inline app/controllers/images_controller.rb]: Remove commented code

File: config/initializers/simple_form.rb [hunk 9]
```diff
@@ -14,7 +15,7 @@
   # stack. The options given below are used to wrap the
   # whole input.
   config.wrappers :default, class: :input,
-    hint_class: :field_with_hint, error_class: :field_with_errors, valid_class: :field_without_errors do |b|
+                            hint_class: :field_with_hint, error_class: :field_with_errors, valid_class: :field_without_errors do |b|
```
Comment (abineetds) [inline config/initializers/simple_form.rb]: if you excluded this file from rubocop, are these changes necessary. We generally try not to change files automatically generated by a gem

File: test/models/image_test.rb [hunk 10]
```diff
@@ -21,40 +20,32 @@ def test_valid
...
-
   # take img obj to a wrong state and check if that error is caught
   # this is prefered in appfolio
 
   # testing style: cause that error and then assert if the expected error message is caught
   def test_is_url_filled
-    @img.image_uri = ""
-    refute_predicate(@img,  :valid?)
+    @img.image_uri = ''
+    assert_not_predicate(@img, :valid?)
```
Comment (abineetds) [inline test/models/image_test.rb]: This is the kind of thing that you might not notice that rubocop might change

Comment (adiengineer) [reply test/models/image_test.rb]: Will correct style issues myself from now on

File: config/initializers/simple_form.rb (hunk 9 above)
Comment (adiengineer) [reply config/initializers/simple_form.rb]: Yeah, will exclude generate files in future and not tinker with them.

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (adiengineer) [reply app/controllers/images_controller.rb]: OK

File: app/controllers/images_controller.rb (hunk 7 above)
Comment (adiengineer) [reply app/controllers/images_controller.rb]: Still not comfortable with the ruby convention of not having brackets!

File: .rubocop.yml (hunk 6 above)
Comment (adiengineer) [reply .rubocop.yml]: Got it.

Comment (adiengineer) [reply .rubocop.yml]: Actually this is funny. I have actually corrected this by */config

Comment (adiengineer) [reply .rubocop.yml]: That is how circleCI passed. Will take a look

File: config/routes.rb [hunk 11]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+
+  resources:images , only: [ :new, :create,:show]
```
[the remark this answers is not in the document]
Comment (adiengineer) [reply config/routes.rb]: Got it

File: app/assets/javascripts/images.js [hunk 12]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
[the remark this answers is not in the document]
Comment (adiengineer) [reply app/assets/javascripts/images.js]: Oh ok. What is that not recommeded? Is there an easy way to automatically filter out auto generated files?

File: .rubocop.yml (hunk 6 above)
Comment (adiengineer) [reply .rubocop.yml]: The last commit removes the full path. Dunno how the old file got pushed up. will commit and push again.

File: config/initializers/simple_form.rb (hunk 9 above)
Comment (adiengineer) [reply config/initializers/simple_form.rb]: Now Excluded this in config

File: config/routes.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (adiengineer) [reply config/routes.rb]: Done

[the remark this answers is not in the document]
Comment (adiengineer) [reply config/routes.rb]: Done

File: test/models/image_test.rb (hunk 2 above)
Comment (adiengineer) [reply test/models/image_test.rb]: Done

File: test/models/image_test.rb (hunk 3 above)
Comment (adiengineer) [reply test/models/image_test.rb]: Yeah but i thought ruby has an implicit return. so the object last assigned/evaluated is returned?

File: test/fixtures/images.yml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
[the remark this answers is not in the document]
Comment (adiengineer) [reply test/fixtures/images.yml]: resolved in a later commit
